22008 staging (#2399)
OpenLiberty/blogs: posts/2022-08-02-separate-stack-trace-22008.adoc

---
layout: post

> title: "Separate stack trace from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8"

Reviewer: ```suggestion
title: "View stack trace separately from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8"
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey

> seo-title: Separate stack trace from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8 - OpenLiberty.io

Reviewer 2: ```suggestion
seo-title: View stack trace separately from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8 - OpenLiberty.io
```

> seo-description: Open Liberty 22.0.0.8 provides a new capability for separating stack trace from messages in logged exceptions by offering new fields for Open Liberty JSON logging and for the Logstash Collector feature. This enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.

Reviewer 2: ```suggestion
seo-description: Open Liberty 22.0.0.8 provides a new capability to view stack trace separately from messages in logged exceptions with new fields for Open Liberty JSON logging and for the Logstash Collector feature. This capability enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.
```

> blog_description: Open Liberty 22.0.0.8 provides a new capability for separating stack trace from messages in logged exceptions by offering new fields for Open Liberty JSON logging and for the Logstash Collector feature. This enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.

Reviewer 2: ```suggestion
blog_description: Open Liberty 22.0.0.8 provides a new capability to view stack trace separately from messages in logged exceptions with new fields for Open Liberty JSON logging and for the Logstash Collector feature. This capability enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = Separate stack trace from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8

Reviewer 2: ```suggestion
= View stack trace separately from messages in logged exceptions and time-based log rollover in Open Liberty 22.0.0.8
```

Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.8 provides a new capability for separating stack trace from messages in logged exceptions by offering new fields for Open Liberty JSON logging and for the Logstash Collector feature. This enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.

Reviewer 2: ```suggestion
Open Liberty 22.0.0.8 provides a new capability to view stack trace separately from messages in logged exceptions with new fields for Open Liberty JSON logging and for the Logstash Collector feature. This capability enables easier-to-read visualizations in downstream log analysis tools. Also new in this release is time-based log rollover, which enables you to specify a time of day for log files to roll over. This release also includes an important security vulnerability (CVE) fix, as well as many notable bug fixes.
```

In link:{url-about}[Open Liberty] 22.0.0.8:

> * <<stack_trace, Stack trace separated from logged messages in logging records>>

Reviewer 2: ```suggestion
* <<stack_trace, View stack trace separately from logged messages in logging records>>
```

* <<log_rollover, Time-based log rollover for Liberty>>
* <<jaxws, jaxws-2.2 updates to CXF 3.4>>
* <<CVEs, Security Vulnerability (CVE) Fixes>>
* <<bugs, Notable bug fixes>>

View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22008+label%3A%22release+bug%22[22.0.0.8].

[#run]
== Run your apps using 22.0.0.8

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>22.0.0.8</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[22.0.0.8,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#stack_trace]

> == Stack trace separated from logged messages in logging records

Reviewer 2: ```suggestion
== View stack trace separately from logged messages in logging records
```

> The stack trace is now separated from logged messages in logging records so that log analysis tools can present them more clearly. This makes visualizations of the logs by downstream log analysis tools easier to read when you are identifying any issues encountered by the application. Previously, any logging record originating from a Java Logger object that made use of any of the methods that accept a `Throwable` parameter would simply append the stack trace to the existing `message` field. Keeping the `message` field solely for the logged message and having a separate field for the stack trace and exception type enhances the effectiveness of log analysis tools down stream.

Reviewer 2: ```suggestion
The stack trace is now separated from logged messages in logging records so that log analysis tools can present them more clearly. This makes visualizations of the logs by downstream log analysis tools easier to read when you are identifying any issues encountered by the application. Previously, any logging record originating from a Java Logger object that made use of any of the methods that accept a `Throwable` parameter would simply append the stack trace to the existing `message` field. Keeping the `message` field solely for the logged message and having a separate field for the stack trace and exception type enhances the effectiveness of downstream log analysis tools.
```

Java's link:https://docs.oracle.com/javase/8/docs/api/java/util/logging/package-summary.html[Logging API] provides methods that allow you to include a `Throwable` as a parameter. When the `Throwable` object is used, Open Liberty's JSON logging provides two new fields, `ibm_stackTrace` and `ibm_exceptionName`, and the Open Liberty link:{url-prefix}/docs/latest/reference/feature/logstashCollector-1.0.html[Logstash Collector] feature provides two new fields, `stackTrace` and `exceptionName`. The stack trace fields present only the stack trace of the `Throwable` object. The exception name fields present the type of exception of the `Throwable` object.

*Example:*

> For the example we will only be showing the message log record for Open Liberty's JSON logging.

Reviewer 2: ```suggestion
For the example, we will only show the message log record for Open Liberty's JSON logging.
```

|link:https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-22476[CVE-2022-22476]
|5
|Identity spoofing
|17.0.0.3 - 22.0.0.7
|Affects the link:{url-prefix}/docs/latest/reference/feature/appSecurity-1.0.html[App Security 1.0], link:{url-prefix}/docs/latest/reference/feature/appSecurity-2.0.html[App Security 2.0], link:{url-prefix}/docs/latest/reference/feature/appSecurity-3.0.html[App Security 3.0] and link:{url-prefix}/docs/latest/reference/feature/appSecurity-4.0.html[App Security 4.0] features
|===

For a list of past security vulnerability fixes, reference the link:{url-prefix}/docs/latest/security-vulnerabilities.html[Security vulnerability (CVE) list].

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A22008+label%3A%22release+bug%22[full list of bugs fixed in 22.0.0.8].

* link:https://github.com/OpenLiberty/open-liberty/issues/21526[UI generated by openapi-3.1 feature doesn't show the link specific endpoints]
+

> An issue arose when you clicked an endpoint in the UI that is generated by the `openapi-3.1 feature`, the clicked endpoint is expanded however the URL of the browser is not changed. As the result, you could not pass a link to a specific endpoint to others for reference. The URL of the browser is expected to change to show the link to the clicked endpoint.

Reviewer 2: ```suggestion
An issue arose that when you clicked an endpoint in the UI that is generated by the `openapi-3.1 feature`, the clicked endpoint is expanded but the URL of the browser is not changed. As the result, you could not pass a link to a specific endpoint to others for reference. The URL of the browser is expected to change to show the link to the clicked endpoint.
```

+
This issue has since been reviewed and subsequently the affected features list have been updated once the UI is shared with the `mpOpenApi` features. This did previously work for openapi-3.1 but had been broken by link:https://github.com/OpenLiberty/open-liberty/pull/19535[#19535] in 22.0.0.2. This issue has now been resolved. 

* link:https://github.com/OpenLiberty/open-liberty/issues/21214[Server start fails when directory has spaces]
+ 

> On Windows operating system, when the server working directory is specified using environment variable SERVER_WORKING_DIR, and when a space is present in the name of the server working directory, an attempt to start the server using the server script will hang and the server will not start. 

Reviewer 2: ```suggestion
On Windows operating system, when the server working directory is specified using the `SERVER_WORKING_DIR` environment variable and a space is present in the name of the server working directory, an attempt to start the server using the server script will hang and the server will not start. 
```

* link:https://github.com/OpenLiberty/open-liberty/issues/21601[Port MYFACES-4432 to JSF 2.3 and Faces 3.0 (Resolve request object in facelets)]
+
The issue occurred when using the `"@FacesConfig"` annotation in a JSF 2.3 or Faces 3.0 application, the `#{request}` object fails to resolve. An empty string is returned instead. 
+
This issue has now been closed as completed in link:https://github.com/OpenLiberty/open-liberty/pull/21599[#21599]

* link:https://github.com/OpenLiberty/open-liberty/issues/21615[EJB persistent timers that were deferred during app start do not run when app finishes starting]
+
EJB Persistent Timers that get deferred due to the application not being started yet do not run after the application finishes starting.
+
Steps to reproduce are to have a large, slowly-starting application with a number of persistent EJB timers around from a previous run. Start the server and if timing is right, the timers will never start running.
+

> The expected behavior is for EJB persistent timers that were deferred due to an unavailable application, should run after the application becomes available again.

Reviewer 2: ```suggestion
The expected behavior is for EJB persistent timers that were deferred due to an unavailable application should run after the application becomes available again.
```
